File: building-asset-management/backend/api/asset.go
package v1

import (
	"net/http"
	"strconv"

	"building-asset-management/internal/model"
	"building-asset-management/internal/service"
	"building-asset-management/pkg/response"

	"github.com/gin-gonic/gin"
)

type AssetAPI struct {
	assetService *service.AssetService
}

func NewAssetAPI() *AssetAPI {
	return &AssetAPI{
		assetService: service.NewAssetService(),
	}
}

// Asset CRUD operations

// GetAssets 获取资产列表
func (a *AssetAPI) GetAssets(c *gin.Context) {
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "10"))
	name := c.Query("name")
	assetType := c.Query("type")
	status := c.Query("status")

	assets, total, err := a.assetService.GetAssets(page, pageSize, name, assetType, status)
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "获取资产列表失败", err.Error())
		return
	}

	response.Success(c, gin.H{
		"list":      assets,
		"total":     total,
		"page":      page,
		"page_size": pageSize,
	})
}

// GetAsset 获取资产详情
func (a *AssetAPI) GetAsset(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 64)
	if err != nil {
		response.Error(c, http.StatusBadRequest, "无效的资产ID", err.Error())
		return
	}

	asset, err := a.assetService.GetAssetByID(uint(id))
	if err != nil {
		response.Error(c, http.StatusNotFound, "资产不存在", err.Error())
		return
	}

	response.Success(c, asset)
}

// CreateAsset 创建资产
func (a *AssetAPI) CreateAsset(c *gin.Context) {
	var req model.Asset
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Error(c, http.StatusBadRequest, "请求参数错误", err.Error())
		return
	}

	userID := c.GetUint("userID")
	req.CreatedBy = userID

	asset, err := a.assetService.CreateAsset(&req)
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "创建资产失败", err.Error())
		return
	}

	response.Success(c, asset)
}

// UpdateAsset 更新资产
func (a *AssetAPI) UpdateAsset(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 64)
	if err != nil {
		response.Error(c, http.StatusBadRequest, "无效的资产ID", err.Error())
		return
	}

	var req model.Asset
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Error(c, http.StatusBadRequest, "请求参数错误", err.Error())
		return
	}

	userID := c.GetUint("userID")
	req.UpdatedBy = userID

	asset, err := a.assetService.UpdateAsset(uint(id), &req)
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "更新资产失败", err.Error())
		return
	}

	response.Success(c, asset)
}

// DeleteAsset 删除资产
func (a *AssetAPI) DeleteAsset(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 64)
	if err != nil {
		response.Error(c, http.StatusBadRequest, "无效的资产ID", err.Error())
		return
	}

	if err := a.assetService.DeleteAsset(uint(id)); err != nil {
		response.Error(c, http.StatusInternalServerError, "删除资产失败", err.Error())
		return
	}

	response.Success(c, nil)
}

// Building CRUD operations

// GetBuildings 获取建筑列表
func (a *AssetAPI) GetBuildings(c *gin.Context) {
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "10"))
	assetID, _ := strconv.ParseUint(c.Query("asset_id"), 10, 64)
	name := c.Query("name")

	buildings, total, err := a.assetService.GetBuildings(page, pageSize, uint(assetID), name)
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "获取建筑列表失败", err.Error())
		return
	}

	response.Success(c, gin.H{
		"list":      buildings,
		"total":     total,
		"page":      page,
		"page_size": pageSize,
	})
}

// GetBuilding 获取建筑详情
func (a *AssetAPI) GetBuilding(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 64)
	if err != nil {
		response.Error(c, http.StatusBadRequest, "无效的建筑ID", err.Error())
		return
	}

	building, err := a.assetService.GetBuildingByID(uint(id))
	if err != nil {
		response.Error(c, http.StatusNotFound, "建筑不存在", err.Error())
		return
	}

	response.Success(c, building)
}

// CreateBuilding 创建建筑
func (a *AssetAPI) CreateBuilding(c *gin.Context) {
	var req model.Building
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Error(c, http.StatusBadRequest, "请求参数错误", err.Error())
		return
	}

	userID := c.GetUint("userID")
	req.CreatedBy = userID

	building, err := a.assetService.CreateBuilding(&req)
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "创建建筑失败", err.Error())
		return
	}

	response.Success(c, building)
}

// UpdateBuilding 更新建筑
func (a *AssetAPI) UpdateBuilding(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 64)
	if err != nil {
		response.Error(c, http.StatusBadRequest, "无效的建筑ID", err.Error())
		return
	}

	var req model.Building
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Error(c, http.StatusBadRequest, "请求参数错误", err.Error())
		return
	}

	userID := c.GetUint("userID")
	req.UpdatedBy = userID

	building, err := a.assetService.UpdateBuilding(uint(id), &req)
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "更新建筑失败", err.Error())
		return
	}

	response.Success(c, building)
}

// DeleteBuilding 删除建筑
func (a *AssetAPI) DeleteBuilding(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 64)
	if err != nil {
		response.Error(c, http.StatusBadRequest, "无效的建筑ID", err.Error())
		return
	}

	if err := a.assetService.DeleteBuilding(uint(id)); err != nil {
		response.Error(c, http.StatusInternalServerError, "删除建筑失败", err.Error())
		return
	}

	response.Success(c, nil)
}

// Floor CRUD operations

// GetFloors 获取楼层列表
func (a *AssetAPI) GetFloors(c *gin.Context) {
	buildingID, _ := strconv.ParseUint(c.Query("building_id"), 10, 64)

	floors, err := a.assetService.GetFloorsByBuildingID(uint(buildingID))
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "获取楼层列表失败", err.Error())
		return
	}

	response.Success(c, floors)
}

// CreateFloor 创建楼层
func (a *AssetAPI) CreateFloor(c *gin.Context) {
	var req model.Floor
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Error(c, http.StatusBadRequest, "请求参数错误", err.Error())
		return
	}

	userID := c.GetUint("userID")
	req.CreatedBy = userID

	floor, err := a.assetService.CreateFloor(&req)
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "创建楼层失败", err.Error())
		return
	}

	response.Success(c, floor)
}

// UpdateFloor 更新楼层
func (a *AssetAPI) UpdateFloor(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 64)
	if err != nil {
		response.Error(c, http.StatusBadRequest, "无效的楼层ID", err.Error())
		return
	}

	var req model.Floor
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Error(c, http.StatusBadRequest, "请求参数错误", err.Error())
		return
	}

	userID := c.GetUint("userID")
	req.UpdatedBy = userID

	floor, err := a.assetService.UpdateFloor(uint(id), &req)
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "更新楼层失败", err.Error())
		return
	}

	response.Success(c, floor)
}

// DeleteFloor 删除楼层
func (a *AssetAPI) DeleteFloor(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 64)
	if err != nil {
		response.Error(c, http.StatusBadRequest, "无效的楼层ID", err.Error())
		return
	}

	if err := a.assetService.DeleteFloor(uint(id)); err != nil {
		response.Error(c, http.StatusInternalServerError, "删除楼层失败", err.Error())
		return
	}

	response.Success(c, nil)
}

// Room CRUD operations

// GetRooms 获取房间列表
func (a *AssetAPI) GetRooms(c *gin.Context) {
	floorID, _ := strconv.ParseUint(c.Query("floor_id"), 10, 64)

	rooms, err := a.assetService.GetRoomsByFloorID(uint(floorID))
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "获取房间列表失败", err.Error())
		return
	}

	response.Success(c, rooms)
}

// CreateRoom 创建房间
func (a *AssetAPI) CreateRoom(c *gin.Context) {
	var req model.Room
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Error(c, http.StatusBadRequest, "请求参数错误", err.Error())
		return
	}

	userID := c.GetUint("userID")
	req.CreatedBy = userID

	room, err := a.assetService.CreateRoom(&req)
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "创建房间失败", err.Error())
		return
	}

	response.Success(c, room)
}

// UpdateRoom 更新房间
func (a *AssetAPI) UpdateRoom(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 64)
	if err != nil {
		response.Error(c, http.StatusBadRequest, "无效的房间ID", err.Error())
		return
	}

	var req model.Room
	if err := c.ShouldBindJSON(&req); err != nil {
		response.Error(c, http.StatusBadRequest, "请求参数错误", err.Error())
		return
	}

	userID := c.GetUint("userID")
	req.UpdatedBy = userID

	room, err := a.assetService.UpdateRoom(uint(id), &req)
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "更新房间失败", err.Error())
		return
	}

	response.Success(c, room)
}

// DeleteRoom 删除房间
func (a *AssetAPI) DeleteRoom(c *gin.Context) {
	id, err := strconv.ParseUint(c.Param("id"), 10, 64)
	if err != nil {
		response.Error(c, http.StatusBadRequest, "无效的房间ID", err.Error())
		return
	}

	if err := a.assetService.DeleteRoom(uint(id)); err != nil {
		response.Error(c, http.StatusInternalServerError, "删除房间失败", err.Error())
		return
	}

	response.Success(c, nil)
}

// GetAssetStatistics 获取资产统计数据
func (a *AssetAPI) GetAssetStatistics(c *gin.Context) {
	stats, err := a.assetService.GetAssetStatistics()
	if err != nil {
		response.Error(c, http.StatusInternalServerError, "获取统计数据失败", err.Error())
		return
	}

	response.Success(c, stats)
}

File: frontend/src/services/asset.ts
import { Asset, Building, Floor, Room, AssetQueryParams, BuildingQueryParams, FloorQueryParams, RoomQueryParams, AssetFormData, BuildingFormData, FloorFormData, RoomFormData, AssetStatistics, BuildingStatistics, FloorStatistics, RoomStatistics } from '../types/asset';
import { get, post, put, del } from '../utils/request';
import { ApiResponse, PageData } from '../types';

// 资产相关API
export const assetService = {
  // 获取资产列表
  async getAssets(params?: AssetQueryParams): Promise<ApiResponse<PageData<Asset>>> {
    return await get('/assets', params);
  },

  // 获取资产详情
  async getAsset(id: number): Promise<ApiResponse<Asset>> {
    return await get(`/assets/${id}`);
  },

  // 创建资产
  async createAsset(data: AssetFormData): Promise<ApiResponse<Asset>> {
    return await post('/assets', data);
  },

  // 更新资产
  async updateAsset(id: number, data: AssetFormData): Promise<ApiResponse<Asset>> {
    return await put(`/assets/${id}`, data);
  },

  // 删除资产
  async deleteAsset(id: number): Promise<ApiResponse<void>> {
    return await del(`/assets/${id}`);
  },

  // 获取资产统计
  async getAssetStatistics(): Promise<ApiResponse<AssetStatistics>> {
    return await get('/assets/statistics');
  },

  // 上传资产图片
  async uploadAssetImage(id: number, file: File): Promise<ApiResponse<string>> {
    const formData = new FormData();
    formData.append('file', file);
    return await post(`/assets/${id}/images`, formData);
  },

  // 删除资产图片
  async deleteAssetImage(id: number, imageUrl: string): Promise<ApiResponse<void>> {
    return await del(`/assets/${id}/images`, { imageUrl });
  },
};

// 楼宇相关API
export const buildingService = {
  // 获取楼宇列表
  async getBuildings(params?: BuildingQueryParams): Promise<ApiResponse<PageData<Building>>> {
    return await get('/buildings', params);
  },

  // 获取楼宇详情
  async getBuilding(id: number): Promise<ApiResponse<Building>> {
    return await get(`/buildings/${id}`);
  },

  // 创建楼宇
  async createBuilding(data: BuildingFormData): Promise<ApiResponse<Building>> {
    return await post('/buildings', data);
  },

  // 更新楼宇
  async updateBuilding(id: number, data: BuildingFormData): Promise<ApiResponse<Building>> {
    return await put(`/buildings/${id}`, data);
  },

  // 删除楼宇
  async deleteBuilding(id: number): Promise<ApiResponse<void>> {
    return await del(`/buildings/${id}`);
  },

  // 获取楼宇统计
  async getBuildingStatistics(): Promise<ApiResponse<BuildingStatistics>> {
    return await get('/buildings/statistics');
  },

  // 根据资产ID获取楼宇列表
  async getBuildingsByAsset(assetId: number): Promise<ApiResponse<Building[]>> {
    return await get('/buildings', { assetId });
  },
};

// 楼层相关API
export const floorService = {
  // 获取楼层列表
  async getFloors(params?: FloorQueryParams): Promise<ApiResponse<PageData<Floor>>> {
    return await get('/floors', params);
  },

  // 获取楼层详情
  async getFloor(id: number): Promise<ApiResponse<Floor>> {
    return await get(`/floors/${id}`);
  },

  // 创建楼层
  async createFloor(data: FloorFormData): Promise<ApiResponse<Floor>> {
    return await post('/floors', data);
  },

  // 更新楼层
  async updateFloor(id: number, data: FloorFormData): Promise<ApiResponse<Floor>> {
    return await put(`/floors/${id}`, data);
  },

  // 删除楼层
  async deleteFloor(id: number): Promise<ApiResponse<void>> {
    return await del(`/floors/${id}`);
  },

  // 获取楼层统计
  async getFloorStatistics(): Promise<ApiResponse<FloorStatistics>> {
    return await get('/floors/statistics');
  },

  // 根据楼宇ID获取楼层列表
  async getFloorsByBuilding(buildingId: number): Promise<ApiResponse<Floor[]>> {
    return await get('/floors', { buildingId });
  },

  // 上传楼层平面图
  async uploadFloorPlan(id: number, file: File): Promise<ApiResponse<string>> {
    const formData = new FormData();
    formData.append('file', file);
    return await post(`/floors/${id}/plan`, formData);
  },
};

// 房间相关API
export const roomService = {
  // 获取房间列表
  async getRooms(params?: RoomQueryParams): Promise<ApiResponse<PageData<Room>>> {
    return await get('/rooms', params);
  },

  // 获取房间详情
  async getRoom(id: number): Promise<ApiResponse<Room>> {
    return await get(`/rooms/${id}`);
  },

  // 创建房间
  async createRoom(data: RoomFormData): Promise<ApiResponse<Room>> {
    return await post('/rooms', data);
  },

  // 更新房间
  async updateRoom(id: number, data: RoomFormData): Promise<ApiResponse<Room>> {
    return await put(`/rooms/${id}`, data);
  },

  // 删除房间
  async deleteRoom(id: number): Promise<ApiResponse<void>> {
    return await del(`/rooms/${id}`);
  },

  // 获取房间统计
  async getRoomStatistics(): Promise<ApiResponse<RoomStatistics>> {
    return await get('/rooms/statistics');
  },

  // 根据楼层ID获取房间列表
  async getRoomsByFloor(floorId: number): Promise<ApiResponse<Room[]>> {
    return await get('/rooms', { floorId });
  },

  // 批量更新房间状态
  async batchUpdateRoomStatus(roomIds: number[], status: string): Promise<ApiResponse<void>> {
    return await put('/rooms/batch-status', { roomIds, status });
  },

  // 房间预约
  async bookRoom(roomId: number, data: {
    startDate: string;
    endDate: string;
    purpose: string;
    contactPerson: string;
    contactPhone: string;
  }): Promise<ApiResponse<void>> {
    return await post(`/rooms/${roomId}/book`, data);
  },

  // 取消房间预约
  async cancelRoomBooking(roomId: number, bookingId: number): Promise<ApiResponse<void>> {
    return await del(`/rooms/${roomId}/book/${bookingId}`);
  },
};

// 地图相关API
export const mapService = {
  // 获取地图点位数据
  async getMapPoints(params?: {
    level?: 'asset' | 'building';
    bounds?: [number, number, number, number]; // [west, south, east, north]
    zoom?: number;
  }): Promise<ApiResponse<any[]>> {
    return await get('/map/points', params);
  },

  // 获取区域统计数据
  async getRegionStatistics(regionCode?: string): Promise<ApiResponse<any>> {
    return await get('/map/regions', { code: regionCode });
  },

  // 地理编码
  async geocode(address: string): Promise<ApiResponse<{
    longitude: number;
    latitude: number;
    formatted_address: string;
  }>> {
    return await get('/map/geocode', { address });
  },

  // 逆地理编码
  async reverseGeocode(longitude: number, latitude: number): Promise<ApiResponse<{
    address: string;
    province: string;
    city: string;
    district: string;
  }>> {
    return await get('/map/reverse-geocode', { longitude, latitude });
  },
};

// 导出服务
export { assetService as default };

// 全局搜索
export const searchService = {
  // 全局搜索
  async globalSearch(query: string, type?: 'asset' | 'building' | 'floor' | 'room'): Promise<ApiResponse<{
    assets: Asset[];
    buildings: Building[];
    floors: Floor[];
    rooms: Room[];
  }>> {
    return await get('/search', { query, type });
  },

  // 搜索建议
  async getSearchSuggestions(query: string): Promise<ApiResponse<string[]>> {
    return await get('/search/suggestions', { query });
  },
};